File: 22. Generate Parentheses/main.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        s = []

        def backtrack(left: int, right: int):
            if left == 0 and right == 0:
                result.append("".join(s))
                return

            if left > 0:
                s.append("(")
                backtrack(left - 1, right)
                s.pop()

            if right > left:
                s.append(")")
                backtrack(left, right - 1)
                s.pop()

        backtrack(n, n)

        return result


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def backtrack(left: int, right: int, s=""):
            if left == 0 and right == 0:
                result.append(s)
                return

            if left > 0:
                backtrack(left - 1, right, s + "(")

            if right > left:
                backtrack(left, right - 1, s + ")")

        result = []
        backtrack(n, n, "")

        return result

File: 22. Generate Parentheses/main.rs
struct Solution;

impl Solution {
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let mut result = Vec::new();

        Self::backtrack(&mut result, String::new(), n, n);

        result
    }

    fn backtrack(result: &mut Vec<String>, s: String, left: i32, right: i32) {
        if left == 0 && right == 0 {
            result.push(s);
            return;
        }

        if left > 0 {
            Self::backtrack(result, s.clone() + "(", left - 1, right);
        }

        if right > left {
            Self::backtrack(result, s.clone() + ")", left, right - 1);
        }
    }
}

impl Solution {
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let mut result = Vec::new();
        let mut s = String::new();

        Self::backtrack(&mut result, &mut s, n, n);

        result
    }

    fn backtrack(result: &mut Vec<String>, s: &mut String, left: i32, right: i32) {
        if left == 0 && right == 0 {
            result.push(s.clone());
            return;
        }

        if left > 0 {
            s.push('(');
            Self::backtrack(result, s, left - 1, right);
            s.pop();
        }

        if right > left {
            s.push(')');
            Self::backtrack(result, s, left, right - 1);
            s.pop();
        }
    }
}
